### PR TITLE
Alert queue optimizations - v1

### PR DESCRIPTION
**Pull request by jufajardini**
Related to (and building upon work from) https://github.com/OISF/suricata/pull/7211

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: https://redmine.openinfosecfoundation.org/issues/4943

Describe changes:
- move pre-processing of the alert queue (PacketAlert) to the `DetectEngineThreadCtx`. This queue grows dynamically as needed. This means appending and sorting the alerts queue from the Detect Engine Thread context, and passing this queue to `PacketAlertFinalize` to apply thresholding, action and alert handling before storing the final PacketAlerts queue in the Packet structure, which will also enforce packet_alert_max limits as defined in our configuration (or default values)
- remove PacketAlert functions that become obsolete with this patch
- add stats about discarded alerts (when we reach packet max alert and the matched signature has a lower priority) and suppressed alerts (when the rule matches, but it's a `noalert` rule or otherwise doesn't reach the threshold to be appended)

Leaving this as draft because I don't know if it will be considered a good idea to be addressing so many different things in a single PR, but local unittests and respective SV tests pass, and commit history is somewhat ready, too, if approved.

Once this work gets approved, then we can see if more work needs to be done in #7212 

suricata-verify-pr: 808
https://github.com/OISF/suricata-verify/pull/808

### CODE REVIEW

File: src/detect.c [hunk 1]
```diff
@@ -828,12 +833,17 @@ static DetectRunScratchpad DetectRunSetup(
 
 #ifdef UNITTESTS
     p->alerts.cnt = 0;
+    p->alerts.discarded = 0;
+    p->alerts.suppressed = 0;
```
Comment (jufajardini) [inline src/detect.c:837]: Wasn't sure if these should stay here or go with the others (outside unittests definition)

File: src/detect.c [hunk 2]
```diff
@@ -1631,7 +1728,7 @@ static void DetectRunFrames(ThreadVars *tv, DetectEngineCtx *de_ctx, DetectEngin
                     det_ctx->frame_id = frame->id;
                     SCLogDebug(
                             "%p/%" PRIi64 " sig %u (%u) matched", frame, frame->id, s->id, s->num);
-                    PacketAlertAppend(det_ctx, s, p, 0, alert_flags); // TODO tx id frame field
```
Comment (jufajardini) [inline src/detect.c:1634]: Is this TODO about tx_id frame field still important? mentioning so it won't get lost...

File: src/detect.c [hunk 3]
```diff
@@ -1001,6 +1017,72 @@ static int RuleMatchCandidateTxArrayExpand(DetectEngineThreadCtx *det_ctx, const
...
+
+/** \internal
+ */
+static void AlertQueueExpand(DetectEngineThreadCtx *det_ctx)
+{
+    det_ctx->alert_queue_capacity *= 2;
+    void *tmp_queue = SCRealloc(
+            det_ctx->alert_queue, (size_t)(sizeof(PacketAlert) * det_ctx->alert_queue_capacity));
+    if (tmp_queue == NULL) {
+        return;
```
Comment (jufajardini) [inline src/detect.c:1047]: Should we throw some error here?

**Review by jufajardini — COMMENTED** [review]
Added a few inline questions about things I wasn't sure.

File: src/detect-engine-alert.c [hunk 4]
```diff
@@ -316,36 +238,42 @@ void PacketAlertFinalize(DetectEngineCtx *de_ctx, DetectEngineThreadCtx *det_ctx
...
         if (res == 0 || res == 2 || (s->flags & SIG_FLAG_NOALERT)) {
-            PacketAlertRemove(p, i);
-
-            if (p->alerts.cnt == 0)
+            /* we will not copy this to the AlertQueue */
+            p->alerts.suppressed++;
+        } else if (p->alerts.cnt < packet_alert_max) {
+            memcpy(&p->alerts.alerts[p->alerts.cnt], &det_ctx->alert_queue[i], sizeof(PacketAlert));
+            SCLogDebug("Appending sid %" PRIu32 " alert to Packet::alerts at pos %u", s->id, i);
+            if (has_pass_action) {
```
Comment (victorjulien) [inline src/detect-engine-alert.c:269]: could we just do `if(PacketTestAction(p, ACTION_PASS)) {` here?

File: src/detect.c (hunk 1 above)
Comment (victorjulien) [reply src/detect.c:837]: are they reset elsewhere as well?

File: src/detect.c [hunk 5]
```diff
@@ -1001,6 +1017,72 @@ static int RuleMatchCandidateTxArrayExpand(DetectEngineThreadCtx *det_ctx, const
...
+    }
+    det_ctx->alert_queue = tmp_queue;
+    SCLogDebug("Alert queue size doubled. %u, bytes: %" PRIuMAX "", det_ctx->alert_queue_capacity,
+            (uintmax_t)(sizeof(PacketAlert) * det_ctx->alert_queue_capacity));
+}
+
+/**
+ * \brief Append signature to local packet alert queue for later preprocessing
+ */
+void DetectEngineThreadCtxAlertQueueAppend(DetectEngineThreadCtx *det_ctx, const Signature *s,
```
Comment (victorjulien) [inline src/detect.c:1057]: Just `AlertQueueAppend`?

File: src/detect.c [hunk 6]
```diff
@@ -1019,6 +1101,21 @@ DetectRunTxSortHelper(const void *a, const void *b)
         return s0->id > s1->id ? 1 : -1;
 }
 
+/** \internal
+ * \brief sort helper for sorting alerts by priority
+ *
+ * The id field is set from Signature::num, so we have higher priority (lower ids) first
+ */
+static int AlertQueueSortHelper(const void *a, const void *b)
```
Comment (victorjulien) [inline src/detect.c:1109]: should maybe make sure that if they are equal we sort by PacketAlert::txid?

File: src/detect.c (hunk 2 above)
Comment (victorjulien) [reply src/detect.c:1634]: ya I think so

**Review by victorjulien — CHANGES_REQUESTED** [review]
Looks pretty good already. Much simpler.

File: src/detect.c (hunk 6 above)
Comment (jufajardini) [reply src/detect.c:1109]: okey dokey!

File: src/detect.c (hunk 2 above)
Comment (jufajardini) [reply src/detect.c:1634]: Can it be made into a ticket?

File: src/decode.c [hunk 7]
```diff
@@ -74,6 +74,26 @@ extern bool stats_decoder_events;
 extern const char *stats_decoder_events_prefix;
 extern bool stats_stream_events;
 uint8_t decoder_max_layers = PKT_DEFAULT_MAX_DECODED_LAYERS;
+uint16_t packet_alert_max = PACKET_ALERT_MAX;
+
+/**
+ * \brief Initialize PacketAlerts with dynamic alerts array size
+ *
+ */
+PacketAlert *CreatePacketAlert(void)
```
[the remark this answers is not in the document]
Comment (jufajardini) [reply src/decode.c:83]: thanks. when work is lost between changes x_x

File: src/detect.c (hunk 1 above)
Comment (jufajardini) [reply src/detect.c:837]: In `PACKET_RELEASE_REFS(p)`

File: src/detect-engine-alert.c (hunk 4 above)
Comment (jufajardini) [reply src/detect-engine-alert.c:269]: Likely. Sometimes, it's hard to see simple things xD I'll try it out.

File: src/detect.c (hunk 3 above)
Comment (jufajardini) [reply src/detect.c:1047]: I think we should, I was thinking something like what we have in https://github.com/OISF/suricata/blob/master/src/detect.c#L992

Comment (jufajardini) [reply src/detect.c:1047]: Decided to go with a warning, for now.

File: src/detect.c (hunk 2 above)
Comment (jufajardini) [reply src/detect.c:1634]: left a comment around the same place, for now

**Comment by suricata-qa** [comment]
ERROR:
 

ERROR: QA failed on tlpw1_files_sha256.
  

Pipeline 7137

**Comment by jufajardini** [comment]
Feedback incorporated in: https://github.com/OISF/suricata/pull/7290

**Pull request closed without merging**